tuta-sdk/rust/src/crypto/entity_facade.rs

```diff
@@ -71,62 +75,48 @@ impl EntityFacade {
         };
 
         return if let AssociationType::Aggregation = association_model.association_type {
-            let association_data = entity.get(association_name);
-            if association_model.cardinality == Cardinality::ZeroOrOne && association_data.is_none() {
-                Ok((ElementValue::Null, errors))
-            } else if association_data.is_none() {
-                Err(ApiCallError::InternalSdkError { error_message: format!("Undefined aggregation {model_name}:{association_name}") })
-            } else if association_model.cardinality == Cardinality::Any {
-                let mut aggregate_vec: Vec<ElementValue> = Vec::new();
-                for aggregate in association_data.unwrap().assert_array().into_iter() {
-                    let aggregate_dict = aggregate.assert_dict();
-                    let mut decrypted_aggregate = match self.decrypt_and_map(aggregate_type_model, aggregate_dict, session_key) {
-                        Ok(dec_aggregate) => dec_aggregate,
-                        Err(_) => {
-                            // Decryption errors are tolerated, so instead of panicking, we just
-                            // collect them for further handling
-                            errors.insert(association_name.to_string(), ElementValue::String(format!("Failed to decrypt association {association_name}")));
-                            continue;
+            match (association_data, association_model.cardinality) {
+                (None, Cardinality::ZeroOrOne) => Ok((ElementValue::Null, errors)),
+                (Some(value), Cardinality::Any) => {
+                    let mut aggregate_vec: Vec<ElementValue> = Vec::new();
+                    for aggregate in value.assert_array().into_iter() {
+                        let mut decrypted_aggregate = match self.decrypt_and_map(aggregate_type_model, aggregate.assert_dict(), session_key) {
+                            Ok(dec_aggregate) => dec_aggregate,
+                            Err(error) => return Err(error)
+                        };
+
+                        // Errors should be grouped inside the top-most object, so they should be
+                        // extracted and removed from aggregates
+                        if decrypted_aggregate.get("errors").is_some() {
```

Comment:
Do you know which of recipients is not working?
Also if we are going to pass recipient around we won't know if it's broken or not

Reply:
> Do you know which of recipients is not working? Also if we are going to pass recipient around we won't know if it's broken or not

We can save the index of the recipient, resulting into the following structure on the topmost object:
```json
{
   "errors":{
      "subject":"Failed to decrypt subject",
      "recipients":{
         "recipients_1":{
            "name":"Failed to decrypt name"
         }
      }
   }
}
```

"recipients_1" can  be changed to just the index

Reply:
You can argue both ways, I don't know if it will be more common for us to check errors on mail or errors on the aggregate but we should not lose the index completely in either case. I think it's fine how you did it now, as long as index is preserved